### PR TITLE
Sliding Sync: Add more tracing

### PR DESCRIPTION
**Pull request by MadLittleMods**
Add more tracing. Spawning from looking at a couple traces and wanting a little more info.

Follow-up to github.com/element-hq/synapse/pull/17501

---

The changes in this PR allow you to find slow Sliding Sync traces ignoring the `wait_for_events` time. In Jaeger, you can now filter for the `current_sync_for_user` operation with `RESULT.result=true` indicating that it actually returned non-empty results.

If you want to find traces for your own user, you can use `RESULT.result=true ARG.sync_config.user="@madlittlemods:matrix.org"`

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://element-hq.github.io/synapse/latest/development/contributing_guide.html#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
  - Feel free to credit yourself, by adding a sentence "Contributed by @github_username." or "Contributed by [Your Name]." to the end of the entry.
* [x] [Code style](https://element-hq.github.io/synapse/latest/code_style.html) is correct
  (run the [linters](https://element-hq.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))

### CODE REVIEW

File: synapse/handlers/sliding_sync.py [hunk 1]
```diff
@@ -737,13 +764,20 @@ async def handle_room(room_id: str) -> None:
...
-        return SlidingSyncResult(
+        sliding_sync_result = SlidingSyncResult(
             next_pos=SlidingSyncStreamToken(to_token, connection_position),
             lists=lists,
             rooms=rooms,
             extensions=extensions,
         )
 
+        # Make it easy to find traces for syncs that aren't empty
+        set_tag(SynapseTags.RESULT_PREFIX + "result", bool(sliding_sync_result))
```
Comment (MadLittleMods) [inline synapse/handlers/sliding_sync.py:775]: Good name for this tag? I'm trying to describe `empty` but would rather avoid the negation. `full` doesn't describe this properly.

Willing to use `empty` with a nudge though.

File: synapse/handlers/sliding_sync.py [hunk 2]
```diff
@@ -737,13 +764,20 @@ async def handle_room(room_id: str) -> None:
...
+        set_tag(SynapseTags.FUNC_ARG_PREFIX + "sync_config.user", user_id)
```
[the remark this answers is not in the document]
Comment (MadLittleMods) [reply synapse/handlers/sliding_sync.py:776]: I think the Jaeger UI is actually doing what I expect but presenting it poorly. It seems that search actually behaves as "Find traces with matching spans" so the min/max duration and sorting applies to the relevant span/operations, not the trace itself. Looks like Jaeger has an issue to clarify this -> https://github.com/jaegertracing/jaeger-ui/issues/397#issuecomment-1553395687

File: synapse/handlers/sliding_sync.py [hunk 3]
```diff
@@ -2092,10 +2160,26 @@ async def get_room_sync_data(
...
+                        num_wild_state_keys,
+                    )
+                    set_tag(
+                        SynapseTags.FUNC_ARG_PREFIX + "required_state_lazy",
+                        lazy_load_room_members,
+                    )
+                    set_tag(
+                        SynapseTags.FUNC_ARG_PREFIX + "required_state_other_count",
+                        num_others,
+                    )
```
Comment (erikjohnston) [inline synapse/handlers/sliding_sync.py:2181]: I think we should just log these to the spans (using `log_kv`), tags are more meant to be searchable things that are limited in scope.

Comment (MadLittleMods) [reply synapse/handlers/sliding_sync.py:2181]: I get the comparison to the prometheus metrics where we do have to worry about cardinality of values (be mindful of limiting th e number possible values).

But we already don't care about this for tracing spans given we set the MXID in `authenticated_entity`, the full `http.url` with specific tokens, `request_id` with unique values, etc.

I can update it if this is something we are going to start caring about. There are lots of other places to update where we tag counts as well.

I tried finding some references in the OpenTracing/OpenTelemetry docs but I didn't really see anything conclusive for tracing span tags/attributes:

References:

 - https://medium.com/@emmanuel.courreges/jaeger-tracing-and-opentelemetry-usage-guidelines-cdc1fcf48415
 - https://opentelemetry.io/docs/concepts/signals/traces/#when-to-use-span-events-versus-span-attributes
 - https://github.com/opentracing/specification/blob/master/specification.md#note-required-formats-for-injection-and-extraction

Comment (erikjohnston) [reply synapse/handlers/sliding_sync.py:2181]: Yeah, fair enough. Though I think it probably would be easier better to not tag these counts? Then its just a single `log_kv({ ... })` call.

Comment (MadLittleMods) [reply synapse/handlers/sliding_sync.py:2181]: Not exactly *easier* given we currently set `required_state_wildcard`/`required_state_wildcard_event_type` in other blocks. But we can always pull the necessary booleans/counts out further and use `log_kv({ ... })` if you prefer.

**Review by erikjohnston — APPROVED** [review]

**Comment by MadLittleMods** [comment]
Thanks for the review @erikjohnston 🐋

**Pull request merged**